lib/smart_proxy_reports/spooled_http_client.rb

```diff
@@ -52,37 +52,44 @@ def stop
       @worker.join
     end
 
+    def connect(http, factory, type, endpoint)
+      processed = 0
+      Dir.glob(spool_path("todo", type + "-*")) do |filename|
+        basename = File.basename(filename)
+        logger.debug "Sending report #{basename}"
+        begin
+          post = factory.create_post(endpoint, File.read(filename))
+          response = http.request(post)
+          logger.info "Report #{basename} sent with HTTP response #{response.code}"
+          logger.debug { "Response body: #{response.body}" }
+          if response.code.start_with?("2")
+            if Proxy::Reports::Plugin.settings.keep_reports
+              spool_move("todo", "done", basename)
+            else
+              FileUtils.rm_f spool_path("todo", basename)
+            end
+          else
+            logger.debug { "Moving failed report to 'fail' spool directory" }
+            spool_move("todo", "done", basename)
+          end
+          processed += 1
+        rescue StandardError => e
+          logger.warn "Unable to send #{basename}, will try on next request: #{e}", e
+          raise if ENV["RACK_ENV"] == "test"
+        end
+      end
+      processed
+    end
+
     def process
       processed = 0
       client = ::Proxy::HttpRequest::ForemanRequest.new
       factory = client.request_factory
       # send all files via a single persistent HTTP connection
       logger.debug "Opening HTTP connection to Foreman"
       client.http.start do |http|
-        Dir.glob(spool_path("todo", "*")) do |filename|
-          basename = File.basename(filename)
-          logger.debug "Sending report #{basename}"
-          begin
-            post = factory.create_post("/api/v2/host_reports", File.read(filename))
-            response = http.request(post)
-            logger.info "Report #{basename} sent with HTTP response #{response.code}"
-            logger.debug { "Response body: #{response.body}" }
-            if response.code.start_with?("2")
-              if Proxy::Reports::Plugin.settings.keep_reports
-                spool_move("todo", "done", basename)
-              else
-                FileUtils.rm_f spool_path("todo", basename)
-              end
-            else
-              logger.debug { "Moving failed report to 'fail' spool directory" }
-              spool_move("todo", "done", basename)
-            end
-            processed += 1
-          rescue StandardError => e
-            logger.warn "Unable to send #{basename}, will try on next request: #{e}", e
-            raise if ENV["RACK_ENV"] == "test"
-          end
-        end
+        processed += connect(http, factory, "facts", "/api/v2/hosts/facts")
+        processed += connect(http, factory, "reports", "/api/v2/host_reports")
```

Comment:
It does not work, you store files with "report" prefix and then you search by "reports" prefix. Singluar vs plural.

```
$ find /tmp/foreman-reports-spool/
/tmp/foreman-reports-spool/
/tmp/foreman-reports-spool/fail
/tmp/foreman-reports-spool/done
/tmp/foreman-reports-spool/todo
/tmp/foreman-reports-spool/todo/report-1b7fa3c9dafe4cbd8c745ec23a84f6d4
/tmp/foreman-reports-spool/todo/report-89ad4b86147943d18080a878038918d2
/tmp/foreman-reports-spool/temp
```

To test this, simply delete the host from foreman and upload any Ansible fixture with facts, Foreman will get facts, create a new host and attach the report to it.

Reply:
You are right, unit tests were also incosistent. It should be fixed by now.